### ISSUE TITLE
List going down

### ISSUE THREAD

**Issue by SBoudrias** [yeoman/yeoman-generator-list#19]
The list is still going down every few weeks. (see prev #17).

For some reason, it looks like our data is getting corrupted and then the server will fail until we manually restart it.

Unlike our previous issue, now the server is going to run fine for multiple weeks. Then _something happen_ and the server will fail until restart. Previously the list would go up and down every hours and come back up on it's own.

I checked the log and nothing relevant was displayed. Chances are the actual bug breaking our data happened previously either silently or is hidden under all the service 500 errors.

I'd like to have more information, but that's all I have for now.

**Comment by SBoudrias** [comment]
The list is still going aggressively down. Also, a new issue starting yesterday is that restarting the Heroku app doesn't fix our problem. It feels like if Heroku is killing our container.

One possible solution is our app taking up too much memory. This would make sense as we're launching hundreds of HTTP requests at the same time when fetching NPM. Maybe a good first step would be to reduce the # of concurrent request we're making.

@eddiemonge Can you report on your finding when you investigated that issue? (I'm interested into what you found and how you found it)

**Comment by eddiemonge** [comment]
Storing the generators as a string kept the memory size under 100mb on my machine. I think fetching it from disk also kept it under 100mb. Anytime a http request was opened, memory shot up at least 100mb more and never went back down. Doing all the requests at once pushed the memory from 350-550 depending on phase of the moon based on the ocean (iow completely random). After all the requests finished memory would drop down to 250-350. Doing the requests again repeated the procedure.

**Comment by eddiemonge** [comment]
Another solution would be to do this not as a web server. We could have a cron job compile the list and then push it to some CDN with a one hour expiration time. I kind of like that solution better since the list would be served from a CDN which has to be faster than what we are doing now with Heroku. Since it would be a cron job, the memory usage might reasonably be lower.

**Comment by SBoudrias** [comment]
Okay, I pushed some fix to hopefully lower the memory impact and I also fixed our last version so it'd run correctly.

Let's wait and see how it looks like.

Oh I also reenabled the newrelic tracking so we'll know when we're exceeding memory quota.

**Comment by sindresorhus** [comment]
> Another solution would be to do this not as a web server. We could have a cron job compile the list and then push it to some CDN with a one hour expiration time. I kind of like that solution better since the list would be served from a CDN which has to be faster than what we are doing now with Heroku. Since it would be a cron job, the memory usage might reasonably be lower.

That was actually always the plan. We just never had the chance to do it...

**Comment by addyosmani** [comment]
Sindre is right. I'm +100 for this solution. It had always been put on the backburner but if you have time to implement it, please go for it. Simon: let's see how your fix goes. Thanks both for trying to make the list more reliable.

**Comment by SBoudrias** [comment]
Well, good news. List is still up.

The memory is very slowly creeping up, so we might have a small memory leak somewhere.

When we update the list cache, our response time goes dramatically up. So that's not super good, but it's not going down.

**Comment by gaboesquivel** [comment]
The compile list could be already filtered, so we only serve the generators that match the naming conventions and are not the blacklist..

**Comment by eddiemonge** [comment]
Ok so using anything other than heroku seems to work fine. The current solution is using something (doesn't matter what at this point, could switch to Google Cloud servers or keep it at my digital ocean instance) to push to Google Cloud Storage (but could also be any other CDN).

**Comment by eddiemonge** [comment]
closing this as it has been pretty stable for the past 3 months.

**Comment by SBoudrias** [comment]
@eddiemonge you should document the new server access / deploy process.

**Comment by eddiemonge** [comment]
Check your email. Maybe, not sure if the way I emailed worked.
